pyomo/core/expr/visitor.py

```diff
@@ -885,13 +1190,20 @@ def evaluate_expression(exp, exception=True, constant=False):
         and is caught.
 
     """
+    clear_active = False
     if constant:
         visitor = _EvaluateConstantExpressionVisitor()
     else:
-        visitor = _EvaluationVisitor(exception=exception)
+        if evaluate_expression.visitor_active:
+            visitor = _EvaluationVisitor(exception=exception)
+        else:
+            visitor = evaluate_expression.visitor_cache
+            visitor.exception = exception
+            evaluate_expression.visitor_active = True
+            clear_active = True
```

Comment:
What is the purpose of these changes? Avoid reconstruction of the visitor each time the function is called?

Reply:
Yes - but still make this thread / recursion-safe.

Reply:
Ah. Got it. Thanks.